stdlib/public/runtime/MetadataLookup.cpp

```diff
@@ -472,6 +473,16 @@ static bool sameObjCTypeManglings(Demangle::NodePointer node1,
 }
 #endif
 
+/// Optimization for the case where we need to compare a StringRef and a null terminated C string
+/// Not converting s2 to a StringRef avoids the need to call both strlen and memcmp when non-matching
+/// but equal length
+static bool stringRefEqualsCString(StringRef s1, const char *s2) {
+  size_t length = s1.size();
+  // It may be possible for s1 to contain embedded NULL characters
+  // so additionally validate that the lengths match
+  return strncmp(s1.data(), s2, length) == 0 && strlen(s2) == length;
```

Comment:
It's a bit unfortunate to have to iterate through s2 twice here. What does performance look like if we use an inline loop here instead of `strncmp` to do the entire comparison in one pass?

Reply:
A naive inline loop seems to be ≈3x slower than platform strncmp in my tests, I suspect due to the optimized ARM assembly implementations in `llvm-project/libc/AOR_v20.02/string/aarch64` + the compiler's optimization of built-ins, but my assembly skills are a bit rusty.

I tried something like this with `-O3`:

```inline static bool testcmp(const char *left, const char *right, size_t n) {
	for (int i=0; i < n; i++) {
		char lc = left[i];
		char rc = right[i];
		
		if (lc != rc || rc == '\0') {
			return false;
		}
	}
	return true;
}
```

Reply:
In that case, it looks good as is. Thanks for trying it out!

Reply:
No worries -- at this level of the code, I think you could only do better if you already knew the length of `s2`, or knew that `s1` contained no embedded `NUL` characters.